### PR TITLE
config: add config file support.

### PR DESCRIPTION
**Pull request by coocood**
Support load configuration from a toml file, and override by arguments.

And refactored `main.go`.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @zimulala @tiancaiamao PTAL

File: tidb-server/main.go [hunk 1]
```diff
@@ -89,163 +139,74 @@ var (
...
+	printInfo()
+	createStoreAndDomain()
+	setupBinlogClient()
+	createServer()
+	setupSignalHandler()
+	setupMetrics()
+	runServer()
+
+	dom.Close()
+	os.Exit(0)
```
Comment (tiancaiamao) [inline tidb-server/main.go]: `storage.Close()` is not called?

File: tidb-server/main.go [hunk 2]
```diff
@@ -89,163 +139,74 @@ var (
...
 
-	if *enablePS {
-		perfschema.EnablePerfSchema()
-	}
-	privileges.Enable = *enablePrivilege
-	privileges.SkipWithGrant = *skipGrantTable
-	if *binlogSocket != "" {
-		createBinlogClient()
-	}
+func registerStores() {
```
Comment (tiancaiamao) [inline tidb-server/main.go]: I suggest register store in `func init()`, some tests may not call main() but they need mocktikv.

File: tidb-server/main.go [hunk 3]
```diff
@@ -89,163 +139,74 @@ var (
...
 )
 
-func main() {
-	tidb.RegisterLocalStore("boltdb", boltdb.Driver{})
-	tidb.RegisterStore("tikv", tikv.Driver{})
-	tidb.RegisterStore("mocktikv", tikv.MockDriver{})
-
-	runtime.GOMAXPROCS(runtime.NumCPU())
+var (
+	cfg     *config.Config
```
Comment (tiancaiamao) [inline tidb-server/main.go]: This is actually a pointer to gloablConfig?

File: tidb-server/main.go [hunk 4]
```diff
@@ -311,3 +272,221 @@ func flagBoolean(name string, defaultVal bool, usage string) *bool {
...
+
+	pushMetric(cfg.Status.MetricsAddr, time.Duration(cfg.Status.MetricsInterval)*time.Second)
+}
+
+func runServer() {
+	if err := svr.Run(); err != nil {
+		log.Error(err)
+	}
+	if cfg.XProtocol.XServer {
+		if err := xsvr.Run(); err != nil {
```
Comment (tiancaiamao) [inline tidb-server/main.go]: `srv.Run()` will not return, so it won't get to `xsvr.Run()` normally...

File: tidb-server/main.go (hunk 2 above)
Comment (coocood) [reply tidb-server/main.go]: init in which package?

File: tidb-server/main.go (hunk 3 above)
Comment (coocood) [reply tidb-server/main.go]: Yes.

File: server/server.go [hunk 5]
```diff
@@ -128,7 +129,7 @@ func (s *Server) newConn(conn net.Conn) *clientConn {
 }
 
 func (s *Server) skipAuth() bool {
```
[the remark this answers is not in the document]
Comment (coocood) [reply server/server.go]: in `readOptionalSSLRequestAndHandshakeResponse` method

File: tidb-server/main.go (hunk 4 above)
Comment (coocood) [reply tidb-server/main.go]: Yes, but it's the same as before, we not going to handle this issue in this PR.

Comment (tiancaiamao) [reply tidb-server/main.go]: okay @wentaoxu

File: config/config_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+type testConfigSuite struct{}
+
+func TestT(t *testing.T) {
+	CustomVerboseFlag = true
+	TestingT(t)
+}
+
+func (s *testConfigSuite) TestConfig(c *C) {
+
```
Comment (zimulala) [inline config/config_test.go]: Remove this blank line.

File: server/http_status.go [hunk 7]
```diff
@@ -18,6 +18,7 @@ import (
 	"net/http"
 	"sync"
 
+	"fmt"
```
Comment (zimulala) [inline server/http_status.go]: Put this line to line 20.

File: server/server.go [hunk 8]
```diff
@@ -40,6 +40,7 @@ import (
 	// For pprof
 	_ "net/http/pprof"
 
+	"fmt"
```
Comment (zimulala) [inline server/server.go]: Ditto.

File: tidb-server/main.go (hunk 4 above)
Comment (wentaoxu) [reply tidb-server/main.go]: thx, @jackysp have fix it, will be merged with branch xwt/x-protocol

**Comment by coocood** [comment]
@zimulala PTAL

**Comment by zimulala** [comment]
LGTM

**Pull request merged**